### ISSUE TITLE
warning for assignment to temporary

### ISSUE THREAD

**Issue by no111u3** [rust-lang/rust-clippy#4497]
For simple useless code:
```
    fn main() {
        (0, ).0 = 1;
    }
```
Compiler not generate any warnings/errors, but clippy generate one warning:

```
warning: assignment to temporary
 --> src/main.rs:4:5
  |
4 |     (0, ).0 = 1;
  |     ^^^^^^^^^^^
  |
```
I think this warning need to implement in compiler too.

**Comment by sanxiyn** [comment]
FYI, this is `temporary_assignment` lint. rust-lang/rust#53224 is the general issue for upstreaming Clippy lints. (`temporary_assignment` is not included in that issue, but it probably should be?)

**Comment by flip1995** [comment]
The rust issue only looks at correctness lints. `temporary_assignment` is a `complexity` lint. Doing something like this cannot result in wrong code, since you can't use the structure you temporary assigned a value to (The assignment operator `=` always returns `()` in Rust, so `a = b = 1`, will result in `a = ()` and `b = 1`*).

So this is valid rust code, that does not have any influence on the correctness of your program. So this doesn't qualify for a rustc lint, because for such cases Clippy exists.

---
\* This seems that it may cause errors, but if you would try to use `a = ()` as an integer somewhere in your program, this would be caught by the rust type system, so we're good here.

**Comment by sanxiyn** [comment]
I very much disagree in that Rust's standard `unused_variables` lint is not a correctness issue either. It does make sense to prioritize correctness lints though.

**Comment by flip1995** [comment]
I agree, that rust has lints, that aren't really correctness concerns. But on the other hand, the rust compiler still has a pretty strict "no new lints" rule, except, there are _really_ good reasons to add a new lint. That's why Clippy exists.

---
As a site note: `unused_variables` can lead to errors:
```rust
let similar_name1 = 10;
let similar_nameI = 11; // unused_variable

assert_eq!(similar_name1, 11); // panic
```

That is a very constructed example, but it does happen in actual code. (BTW, Clippy has a lint for this: https://rust-lang.github.io/rust-clippy/master/index.html#similar_names)
